crates/walrus-service/src/node.rs

```diff
@@ -2835,53 +2973,100 @@ impl ServiceState for StorageNodeInner {
+            .as_ref()
+            .and_then(|info| info.invalidation_event())
+        {
+            return Err(StoreMetadataError::InvalidBlob(event));
         }
 
-        // Check if encoding type is supported
         let encoding_type = metadata.metadata().encoding_type();
-        if !encoding_type.is_supported() {
-            return Err(StoreMetadataError::UnsupportedEncodingType(encoding_type));
-        }
+        ensure!(
+            encoding_type.is_supported(),
+            StoreMetadataError::UnsupportedEncodingType(encoding_type),
+        );
 
         let encoding_config = self.encoding_config.clone();
-        let verified_metadata_with_id = self
-            .thread_pool
-            .clone()
-            .oneshot(move || metadata.verify(&encoding_config))
-            .map(thread_pool::unwrap_or_resume_panic)
-            .await?;
+        let verified = Arc::new(
+            self.thread_pool
+                .clone()
+                .oneshot(move || metadata.verify(&encoding_config))
+                .map(thread_pool::unwrap_or_resume_panic)
+                .await?,
+        );
 
-        self.storage
-            .put_verified_metadata(&verified_metadata_with_id)
-            .await
-            .context("unable to store metadata")?;
+        let blob_info = self
+            .storage
+            .get_blob_info(&blob_id)
+            .context("could not retrieve blob info")?;
 
-        self.metrics
-            .uploaded_metadata_unencoded_blob_bytes
-            .observe(verified_metadata_with_id.as_ref().unencoded_length() as f64);
-        self.metrics.metadata_stored_total.inc();
+        if blob_info
+            .as_ref()
+            .is_some_and(|info| info.is_registered(self.current_committee_epoch()))
+        {
+            if blob_info
+                .as_ref()
+                .is_some_and(|info| info.is_metadata_stored())
+            {
+                return Ok(false);
+            }
 
-        Ok(true)
+            self.pending_metadata_cache.remove(&blob_id).await;
+
+            self.storage
+                .put_verified_metadata(&verified)
+                .await
+                .context("unable to store metadata")
+                .map_err(StoreMetadataError::Internal)?;
+
+            self.metrics
+                .uploaded_metadata_unencoded_blob_bytes
+                .observe(verified.metadata().unencoded_length() as f64);
+            self.metrics.metadata_stored_total.inc();
+
+            Ok(true)
+        } else {
+            if !intent.is_pending() {
+                return Err(StoreMetadataError::NotCurrentlyRegistered);
+            }
+
+            self.pending_metadata_cache
+                .insert(blob_id, verified)
+                .await
+                .map_err(|_| StoreMetadataError::CacheSaturated)?;
+
+            if self.is_blob_registered(&blob_id)?
+                && let Err(error) = self.flush_pending_metadata(&blob_id).await
```

Comment:
Question: Why is this needed? We already checked if the blob is registered a few lines above.